airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py

```diff
@@ -0,0 +1,388 @@
+        # This method is called once for each record returned from the API to compare the cursor field value in that record with the current state
+        # we then return an updated state object. If this is the first time we run a sync or no state was passed, current_stream_state will be None.
+        latest_record_date_str: str = self.get_field(latest_record, self.cursor_field)
+
+        if current_stream_state and "date" in current_stream_state and latest_record_date_str:
+            # isoparse supports different formats, like:
+            # python iso format:               2021-06-04T00:00:00+03:00
+            # format from transactions record: 2021-06-04T00:00:00+0300
+            # format from balances record:     2021-06-02T00:00:00Z
+            latest_record_date = isoparse(latest_record_date_str)
+            current_parsed_date = isoparse(current_stream_state["date"])
+
+            return {"date": max(current_parsed_date, latest_record_date).isoformat()}
+        else:
+            return {"date": self.start_date.isoformat()}
+
+    def get_last_refreshed_datetime(self, sync_mode):
+        """Get last_refreshed_datetime attribute from API response by running PaypalTransactionStream().read_records()
+        with 'empty' stream_slice (range=0)
+
+        last_refreshed_datetime indicates the maximum available start_date for which API has data.
+        If request start_date > last_refreshed_datetime then API throws an error:
+            "Data for the given start date is not available"
+        """
+        paypal_stream = self.__class__(
+            authenticator=self.authenticator,
+            start_date=self.start_date,
+            end_date=self.start_date,
+            is_sandbox=self.is_sandbox,
+        )
+        stream_slice = {
+            "start_date": self.start_date.isoformat(),
+            "end_date": self.start_date.isoformat(),
+        }
+        list(paypal_stream.read_records(sync_mode=sync_mode, stream_slice=stream_slice))
+        return paypal_stream.last_refreshed_datetime
+
+    def stream_slices(
+        self, sync_mode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, any]]]:
+        """
+        Returns a list of slices for each day (by default) between the start date and end date.
+        The return value is a list of dicts {'start_date': date_string, 'end_date': date_string}.
+        """
+        period = timedelta(**self.stream_slice_period)
+
+        # get last_refreshed_datetime from API response to use as maximum allowed start_date
+        self.last_refreshed_datetime = self.get_last_refreshed_datetime(sync_mode)
+        if self.last_refreshed_datetime:
+            self.logger.info(f"Maximum allowed start_date is {self.last_refreshed_datetime} based on info from API response")
+            self.maximum_allowed_start_date = min(self.last_refreshed_datetime, self.maximum_allowed_start_date)
```

Comment:
maximum_allowed_start_date can be found in API response. So we send 1 'empty' API request (date_range = 0) just before we construct all stream slices.